cheatsheets/File_Upload_Cheat_Sheet.md

```diff
@@ -0,0 +1,163 @@
+
+> This should not be used on its own, as bypassing it is pretty common and easy.
+
+### File Name Sanitization
+
+Multiple threats relate to the file-name, thus, it is essential to ensure that the file-name is stored in a secure fashion.
+
+File-names can endager the system in multiple ways, either by using non acceptable characters, or by using special and restricted filenames, which can be found under the following [MSDN guide](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file?redirectedfrom=MSDN#naming-conventions). For a wider overview on different filesystems and how they treat files, refer to [Wikipedia's Filename page](https://en.wikipedia.org/wiki/Filename).
+
+In order to avoid the above mentioned threat, creating a **random string** as a file-name, _e.g_ generating a UUID/GUID, is essential. If the file-name is required by the business needs, proper input validation should be done for client-side (_e.g._ active content) and back-end side (_e.g._ special files overwrite or creation) attack vectors. File-name length limits should be taken into consideration based on the system storing the files, as each system has its own file name length limit.
+
+### File Content Validation
+
+As mentioned in the [Public File Retrieval](#public-file-retrieval) section, file content can contain unwelcomed data and information.
+
+The File Upload service should allow users to report illegal content, and copyright owners to report abuse.
+
+If there are enough resources, manual file review should be conducted in a sandboxed environment before releasing the files to the public.
+
+Adding some automation to the review could be helpful, which is a harsh process and well studied before its usage. Some services (_e.g._ Virus Total) provide APIs to scan files against well known malicious file hashes. Some frameworks can check and validate the raw content type and validating it against predefined file types, such as in [ASP.NET Drawing Library](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.imaging.imageformat)
+
+### File Storage Location
+
+The location where the files should be stored must be chosen based on security and business requirements. The following points are set by security priority, and are inclusive:
+
+1. Store the files on a **different host**, which allow for complete segragation of duties between the application serving the user, and the host handling file uploads and their storage.
+2. Store the files **outside the webroot**, where only administrative access is allowed.
+3. Store the files **inside the webroot**, and set them in write permissions only.
+   - If read access is required, setting proper controls is a must (_e.g._ internal IP, authorized user, etc.)
+4. Store the files to a **database**, most preferrably as Blobs in a NoSQL database. The database should be chosen based on the application's requirements.
```

Comment:
Why is this considered less secure than storing them in the webroot?

Reply:
After doing additional research, I'll modify this point. There are tons of thoughts onto this that we shouldn't actually dig into. I'll mainly just tell the readers to remember that storing files in a DB is possible.
It is more secure because you will require an SQL configuration issue to happen, which is harder to happen versus a permissions issue on the filesystem. It kind of relieves security issues, but adds burden on the DBA to properly handle the file storage and how to actually separate them from the rest of the DB tables.